### PR TITLE
TotalSegmentatorv2

### PR DESCRIPTION
**Pull request by rbumm**
Please have a first look at this. 

Major changes, not evrything working so far. 
pip installs come as exe files on windows, but what to expect on Linux and MAC ? 
The UI has a conflict. Can you resolve that ? 
Best
Rudolf

### CODE REVIEW

File: TotalSegmentator/TotalSegmentator.py [hunk 1]
```diff
@@ -803,7 +792,7 @@ def process(self, inputVolume, outputSegmentation, fast=True, task=None, subset=
         # Get TotalSegmentator launcher command
         # TotalSegmentator (.py file, without extension) is installed in Python Scripts folder
         import sysconfig
-        totalSegmentatorPath = os.path.join(sysconfig.get_path('scripts'), "TotalSegmentator")
+        totalSegmentatorPath = os.path.join(sysconfig.get_path('scripts'), "TotalSegmentator.exe")
```
Comment (lassoan) [inline TotalSegmentator/TotalSegmentator.py:795]: You don't need to use this launcher. It is just a shortcut for `python.exe -m TotalSegmentator`. Use the Python executable instead (see for example how `pip.exe` is launched by `pip_install` function).

**Review by lassoan — COMMENTED** [review]
Thanks a lot for working on this!
Is there any specific that does not work yet?

**Comment by lassoan** [comment]
There are some merge conflicts, probably because you have not started from the latest main version. Could you rebase your changes on the latest main version of this repository?

**Comment by rbumm** [comment]
I tried to, but could not resolve the small merge conflict. Please help.

**Comment by rbumm** [comment]
I have now resolved the merge conflict.

File: TotalSegmentator/TotalSegmentator.py (hunk 1 above)
Comment (rbumm) [reply TotalSegmentator/TotalSegmentator.py:795]: Please explain.

**Comment by rbumm** [comment]
@lassoan After a pip install TotalSegmentator from this PR extension, I get all kinds of ITK errors. Be prepared to test this in a separate installation.

**Comment by rbumm** [comment]
After installing TotalSegmentator v2, when I use Lung CT Segmenter with lungmask, I get: 

RuntimeError: Exception thrown in SimpleITK ImageFileReader_Execute: D:\a\1\sitk\Code\IO\src\sitkImageReaderBase.cxx:97:
sitk::ERROR: The file "slicer:00000128B3AB8E00#vtkMRMLScalarVolumeNode1" does not exist.
[VTK] Warning: In vtkMRMLSegmentationDisplayNode.cxx, line 295
[VTK] vtkMRMLSegmentationDisplayNode (00000128D4226E50): vtkMRMLSegmentationDisplayNode::GetSegmentDisplayProperties: no display properties are found for segment ID=, return default

**Comment by rbumm** [comment]
Required-by: lungmask, MedPy, nnunet-customized, TotalSegmentator

**Comment by rbumm** [comment]
But lungmask does not work with 2.3.0

**Comment by lassoan** [comment]
SimpleITK requirements:
- https://github.com/JoHof/lungmask/blob/3c9da45f809494368d3679dee87417ad344f670f/requirements.txt#L7 - any
- https://github.com/loli/medpy/blob/66265de8aedf6259feac00b897a22d0cf173d2e2/setup.py#L137 >= 1.1.0
- https://github.com/MIC-DKFZ/nnUNet/blob/de48541317a8019668cfb37a354da241040fe3ce/pyproject.toml#L43 >= 2.2.1
- https://github.com/wasserth/TotalSegmentator/blob/216397e52ec70d098e495793a980527999e6805c/setup.py#L28 - any

These are all reasonable and the overall requirement is `>=2.2.1`. In Slicer-5.4 SimpleITK version is  '2.2.0rc2.dev368', so we would just need to bump it up to 2.2.1 to prevent TotalSegmentator to remove Slicer's SimpleITK. I'll give it a try and if everything seems to work with 2.2.1 then I'll send a pull request to upgrade Slicer's SimpleITK.

Does lungmask work with SimpleITK 2.2.1? What error do you get if you use a more recent SimpleITK version? It may be a very easy fix, as SimpleITK is only used at a few places for trivial operations - see https://github.com/search?q=repo%3AJoHof%2Flungmask+sitk&type=code

**Comment by lassoan** [comment]
@jamesobutler I see that you have updated Python packages many times in the past, including SimpleITK. Would you mind having a look if you could update Slicer's SimpleITK to 2.2.1? The upgrade  would make it much easier for us to run TotalSegmentator v2.

**Comment by jamesobutler** [comment]
Sure I can take a look this evening. The situation with 3D Slicer's SimpleITK is that JC was maintaining some namespace customizations (https://github.com/Slicer/SimpleITK/commits/slicer-v2.2.0-2022-08-30-1c0cf5de) to support the `itk` python package being installed as well.

**Comment by lassoan** [comment]
I see OK, so it may not be so simple then. I've added an issue in the Slicer bugtracker to discuss this further: https://github.com/Slicer/Slicer/issues/7279

**Comment by lassoan** [comment]
@rbumm I've resolved the merge conflict and force-pushed. Make sure to do any further changes (if needed) based on the current content of this branch.

I'll now have a look if I can make the SimpleITK hack work again.

**Comment by rbumm** [comment]
In Windows 11 force installing the latest TS version results in a TotalSegmentator.exe in the script dir. 

image

I can only run TS 2.0.3 from the extension if I use this TotalSegmentator.exe file in the extensions command line.

**Comment by rbumm** [comment]
In addition, I got strange results: 

image

**Comment by lassoan** [comment]
The launcher scripts in the Scripts folder are OK. That's how we launch TotalSegmentator on all platforms. Instead of the command-line interface we could use the Python API directly in the Slicer process, but it is safer to run the segmentation in a separate process.

I also experienced that the "total" task in fast mode tends to leave these huge holes around the heart (extending into the spine). I would recommend to create a separate issue for that. Most likely it is an error in the new model and should be reported to Jakob.

**Comment by lassoan** [comment]
Everything should work well now with the latest changes that I've pushed into this branch.
TotalSegmentator installs fully automatically without even requiring a restart.

You have removed the installLatestDevelopmentCheckBox and I'm completely fine with that, as it is better if we test a new version of TotalSegmentator and if it works then we update the Slicer extension (and when the user upgrades the extension then he will get the updated TotalSegmentator as well). But if you want to allow users to use yet untested TotalSegmentator versions then that's fine for me, too (then just add back the checkbox and related code).

There is just one thing to do before releasing: adding segment information (segment names, codes, colors) for the new segments. Do you think you could you work on this?

FYI, TotalSegmentator segmentation results can be used with the new Colorize volume module (in Sandbox) to create beautiful volume renderings: https://discourse.slicer.org/t/vtk-multivolume-cinematic-volume-rendering/31981/25?u=lassoan

**Comment by lassoan** [comment]
Thanks for the fix! (I've added some comments to explain the changes and did not retest after that).

What do you think about being able to select TotalSegmentator version:
- Should we still support 1.x? (mostly because the fast mode produces a big hole around the heart in 2.0.3) 
- Should we allow using the very latest version? (a bit more work for us, but it gives more freedom to the user) 

Can you work on getting terminology and colors for the new classes? I believe it should be altrady available somewhere (as every software that needs to save in DICOM Segmentation Object needs this) and probably it should be added to TotalSegmentator repository, and Slicer should get the informatíon from there.

**Comment by rbumm** [comment]
>>Should we still support 1.x? (mostly because the fast mode produces a big hole around the heart in 2.0.3)
Yes for backward compatibility

>>Should we allow using the very latest version? (a bit more work for us, but it gives more freedom to the user)
If possible yes

>>Can you work on getting terminology and colors for the new classes? 

I can try but have not found sources yet ...

**Comment by rbumm** [comment]
// Can you work on getting terminology and colors for the new classes?

Where would I add colors?

**Comment by rbumm** [comment]
TS v2 seems to be downloading weights every time it operates - nothing gets buffered - not sure that is the intended way ... Will keep on testing

**Comment by lassoan** [comment]
Colors are defined terminology json files. Two of them are bundled with Slicer core. We could update those (see for example [this pull request](https://github.com/Slicer/Slicer/pull/6728/files)) or we could add a new terminology file to SlicerTotalSegmentator extension.

I noticed the progress bars each time executing TS. Do they indicate downloads? If you suspect yes then you can attach a debugger (https://slicer.readthedocs.io/en/latest/developer_guide/debugging/python.html), add a breakpoint, and see why the model is downloaded again.

**Comment by rbumm** [comment]
After my SNOMED editions, although done very carefully, I get errors

Importing segmentation results...
Failed to deserialize terminology string: Segmentation category and type - Total Segmentator~SCT^123037004^Anatomical Structure~SCT^69748006^Thyroid structure~^^~Anatomic codes - DICOM master list~^^~^^|
Failed to deserialize terminology string: Segmentation category and type - Total Segmentator~SCT^123037004^Anatomical Structure~SCT^29700009^Subclavian artery~SCT^24028007^Right~Anatomic codes - DICOM master list~^^~^^|
Failed to deserialize terminology string: Segmentation category and type - Total Segmentator~SCT^123037004^Anatomical Structure~SCT^85235006^Subclavian artery~SCT^7771000^Left~Anatomic codes - DICOM master list~^^~^^|
Failed to deserialize terminology string: Segmentation category and type - Total Segmentator~SCT^123037004^Anatomical Structure~SCT^65355003^Right common carotid artery~SCT^24028007^Right~Anatomic codes - DICOM master list~^^~^^|
Failed to deserialize terminology string: Segmentation category and type - Total Segmentator~SCT^123037004^Anatomical Structure~SCT^113263003^Left common carotid artery~SCT^7771000^Left~Anatomic codes - DICOM master list~^^~^^|
Failed to deserialize terminology string: Segmentation category and type - Total Segmentator~SCT^123037004^Anatomical Structure~SCT^8887007^Brachiocephalic vein~SCT^7771000^Left~Anatomic codes - DICOM master list~^^~^^|
Failed to deserialize terminology string: Segmentation category and type - Total Segmentator~SCT^123037004^Anatomical Structure~SCT^8887007^Brachiocephalic vein~SCT^24028007^Right~Anatomic codes - DICOM master list~^^~^^|
Failed to deserialize terminology string: Segmentation category and type - Total Segmentator~SCT^123037004^Anatomical Structure~SCT^ 50016007^Rib cartilage~SCT^^Anatomic codes - DICOM master list~^^~^^|
Cleaning up temporary folder...

**Comment by rbumm** [comment]
The downloading only happens in FAST mode. It is certainly not intended. I will create an issue.

**Comment by lassoan** [comment]
Thanks for the updates. 
I'll try to have a look at the terminology/colors definition in Slicer core.

**Comment by lassoan** [comment]
> The downloading only happens in FAST mode. It is certainly not intended. I will create an issue.

@rbumm Have you been able to find out something?

> I also experienced that the "total" task in fast mode tends to leave these huge holes around the heart (extending into the spine). I would recommend to create a separate issue for that. Most likely it is an error in the new model and should be reported to Jakob.

@rbumm Did you end up asking Jakob about it? It means that fast mode is not really usable in TotalSegmentator v2. It is a major issue because those who only have CPU rely on using the fast mode.

**Comment by fedorov** [comment]
@lassoan can you explain the relation between terminology stored in https://github.com/lassoan/SlicerTotalSegmentator/blob/main/TotalSegmentator/Resources/SegmentationCategoryTypeModifier-TotalSegmentator.term.json and the cryptic content in these lines: https://github.com/lassoan/SlicerTotalSegmentator/blob/main/TotalSegmentator/TotalSegmentator.py#L348-L501. 

Are there good reasons not to use JSON file as the one and only place to maintain mapping? I see a number of advantages:
* easier to review and edit
* possible to do automatic checks for consistency
* interoperable with other tools
* no duplication and confusion as to where mapping is performed

We already have `3dSlicerLabel` attribute in that JSON. In the same manner we could introduce something like `Totalsegmentatorv2Label`, and load that JSON into the dictionary in the module code.

**Comment by rbumm** [comment]
He did not react, surry

**Comment by lassoan** [comment]
> the cryptic content in these lines: https://github.com/lassoan/SlicerTotalSegmentator/blob/main/TotalSegmentator/TotalSegmentator.py#L348-L501.

This is how we stored the TotalSegmentator -> SNOMED CT mapping before you created the mapping .csv file. We now use that .csv file.

> Are there good reasons not to use JSON file as the one and only place to maintain mapping?

We have been always using the standard DCMQI terminology JSON files ([DICOM in Slicer core](https://github.com/Slicer/Slicer/blob/main/Modules/Loadable/Terminologies/Resources/SegmentationCategoryTypeModifier-DICOM-Master.json) and [some additional entries for TotalSegmentator](https://github.com/lassoan/SlicerTotalSegmentator/blob/main/TotalSegmentator/Resources/SegmentationCategoryTypeModifier-TotalSegmentator.term.json)). These JSON files are not for mapping but to specify the terminology items, so we still need the mapping .csv file.

**Comment by lassoan** [comment]
I've accidentally closed this pull request. We continue tracking the remaining few tasks here: https://github.com/lassoan/SlicerTotalSegmentator/pull/38

@fedorov it would be great if you could have a look at the mapping .csv and terminology .json files.

**Pull request merged**